src/agent/onefuzz-agent/src/worker.rs

```diff
@@ -159,13 +226,51 @@ impl State<Running> {
                 ctx,
                 work: self.work,
             };
+
+            // Since the agent has exited, drain one more time before returning
+            while let Ok(res) = self.ctx.from_task_to_agent.try_recv() {
+                info!("received message from server_receiver: {:?}", res);
+            }
             Ok(Waited::Done(state))
         } else {
             Ok(Waited::Running(self))
         }
     }
 
-    pub fn kill(&mut self) -> Result<()> {
+    pub async fn kill(&mut self) -> Result<()> {
+        let graceful_shutdown = self.ctx.from_agent_to_task.send(IpcMessageKind::Shutdown);
+
+        for i in 0..90 {
+            match graceful_shutdown {
+                Ok(_) => match self.ctx.child.try_wait() {
+                    Ok(Some(_)) => {
+                        info!("Agent was gracefully shut down");
+                        // Since the agent has exited, drain one more time before returning
+                        while let Ok(res) = self.ctx.from_task_to_agent.try_recv() {
+                            info!("received message from server_receiver: {:?}", res);
+                        }
+                        return Ok(());
+                    }
+                    Ok(None) => {
+                        if i == 89 {
```

Comment:
Why not use the `time::timeout` for the logic of waiting for the process to return like you did in `State<Ready>::run`  ?
If you have to do it this way at least make the end condition dependent on the loop counter. 
``` rust 
    if i == MAX_COUNT - 1 
```